clients/drcachesim/drpt2trace/pt2ir.cpp

```diff
@@ -0,0 +1,395 @@
+    } while (0)
+
+pt2ir_t::pt2ir_t(IN const pt2ir_config_t config)
+    : config_(config)
+    , pt_raw_buffer_(nullptr)
+    , pt_raw_buffer_size_(0)
+    , pt_instr_decoder_(nullptr)
+    , pt_iscache_(nullptr)
+    , pt_sb_session_(nullptr)
+{
+    pt_iscache_ = pt_iscache_alloc(NULL);
+    if (pt_iscache_ == NULL) {
+        ERRMSG("Failed to allocate iscache.\n");
+        exit(1);
+    }
+    pt_sb_session_ = pt_sb_alloc(pt_iscache_);
+    if (pt_sb_session_ == NULL) {
+        ERRMSG("Failed to allocate sb session.\n");
+        pt_iscache_free(pt_iscache_);
+        exit(1);
+    }
+    if (!init()) {
+        ERRMSG("Failed to initialize.\n");
+        pt_iscache_free(pt_iscache_);
+        pt_sb_free(pt_sb_session_);
+        exit(1);
+    }
+}
+
+pt2ir_t::~pt2ir_t()
+{
+    pt_sb_free(pt_sb_session_);
+    pt_iscache_free(pt_iscache_);
+    pt_insn_free_decoder(pt_instr_decoder_);
+    if (pt_raw_buffer_ != NULL) {
+        free(pt_raw_buffer_);
+    }
+}
+
+pt2ir_convert_status_t
+pt2ir_t::convert()
+{
+    /* PT raw data consists of many packets. And PT trace data is surrounded by Packet
+     * Stream Boundary. So, in the outermost loop, this function first finds the PSB. Then
+     * it decodes the trace data. */
+    for (;;) {
+        struct pt_insn insn;
+        memset(&insn, 0, sizeof(insn));
+        int status = 0;
+
+        /* Sync decoder to the first Packet Stream Boundary (PSB) packet, and then decode
+         * instructions. If there is no PSB packet, the decoder will be synced to the end
+         * of the trace. If an error occurs, we will call dx_decoding_error to print the
+         * error information. What are PSB packets? Below answer is quoted from Intel 64
+         * and IA-32 Architectures Software Developer’s Manual 32.1.1.1 Packet Summary
+         * “Packet Stream Boundary (PSB) packets: PSB packets act as ‘heartbeats’ that are
+         * generated at regular intervals (e.g., every 4K trace packet bytes). These
+         * packets allow the packet decoder to find the packet boundaries within the
+         * output data stream; a PSB packet should be the first packet that a decoder
+         * looks for when beginning to decode a trace.”
+         */
+        status = pt_insn_sync_forward(pt_instr_decoder_);
+        if (status < 0) {
+            if (status == -pte_eos)
+                break;
+            dx_decoding_error(status, "sync error", insn.ip);
+            return PT2IR_CONV_SYNC_PACKET_ERROR;
+        }
+
+        /* Decode the raw trace data surround by PSB. */
+        for (;;) {
+            int nextstatus = status;
+            int errcode = 0;
+            struct pt_image *image;
+
+            /* Before starting to decode instructions, we need to handle the event before
+             * the instruction trace. For example, if a mmap2 event happens, we need to
+             * switch the cached image. */
+            while (nextstatus & pts_event_pending) {
```

Comment:
style: Explicit bools (so compare to 0).

Reply:
Done.